src/ini-load.js

```diff
@@ -56,13 +86,83 @@ export function whenContentIniLoad(
     const promises = [];
     resources.forEach((r) => {
       if (!EXCLUDE_INI_LOAD.includes(r.element.tagName)) {
-        promises.push(r.loadedOnce());
+        promises.push(r.element.whenLoaded());
       }
     });
     return Promise.all(promises);
   });
 }
 
+/**
+ * A new way using direct measurement.
+ *
+ * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+ * @param {boolean=} opt_prerenderableOnly signifies if we are in prerender mode.
+ * @return {!Promise}
+ * @visibleForTesting
+ * TODO(#31915): remove visibility
+ */
+export function whenContentIniLoadInOb(elementOrAmpDoc, opt_prerenderableOnly) {
+  const ampdoc = Services.ampdoc(elementOrAmpDoc);
+  return (
+    ampdoc
+      .signals()
+      // First, wait for the `ready-scan` signal. Waiting for each element
+      // individually is too expensive and `ready-scan` will cover most of
+      // the initially parsed elements.
+      .whenSignal(READY_SCAN_SIGNAL)
+      .then(() => {
+        // Filter elements.
```

Comment:
Nit: Can we extract this out of the `ampdoc.signal()…` call chain? It's just very indented.

Reply:
I unindented it a bit. Does it look better?